File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/settings/preference/ProxySettingDialog.java
package jp.hazuki.yuzubrowser.legacy.settings.preference;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;

import jp.hazuki.yuzubrowser.legacy.R;
import jp.hazuki.yuzubrowser.legacy.settings.data.AppData;
import jp.hazuki.yuzubrowser.legacy.settings.preference.common.CustomDialogPreference;
import jp.hazuki.yuzubrowser.legacy.webkit.WebViewProxy;

public class ProxySettingDialog extends CustomDialogPreference {
    private boolean mSaveSettings = true;

    public ProxySettingDialog(Context context) {
        super(context);
    }

    public ProxySettingDialog(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ProxySettingDialog setSaveSettings(boolean save) {
        mSaveSettings = save;
        return this;
    }

    @NonNull
    @Override
    protected CustomDialogFragment crateCustomDialog() {
        return SettingDialog.newInstance(mSaveSettings);
    }

    public static class SettingDialog extends CustomDialogFragment {
        private static final String SAVE = "save";

        public static SettingDialog newInstance(boolean save) {
            SettingDialog dialog = new SettingDialog();
            Bundle bundle = new Bundle();
            bundle.putBoolean(SAVE, save);
            dialog.setArguments(bundle);
            return dialog;
        }

        @NonNull
        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            View view = LayoutInflater.from(getContext()).inflate(R.layout.proxy_dialog, null);
            final CheckBox checkBox = view.findViewById(R.id.checkBox);
            final EditText editText = view.findViewById(R.id.editText);

            checkBox.setChecked(AppData.proxy_set.get());
            editText.setText(AppData.proxy_address.get());

            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

            builder
                    .setView(view)
                    .setTitle(R.string.pref_proxy_settings)
                    .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                        Context context = getContext();
                        boolean enable = checkBox.isChecked();
                        String proxy_address = editText.getText().toString();

                        WebViewProxy.setProxy(context, enable, proxy_address);

                        if (getArguments().getBoolean(SAVE)) {
                            AppData.proxy_set.set(enable);
                            AppData.proxy_address.set(proxy_address);
                            AppData.commit(context, AppData.proxy_set, AppData.proxy_address);
                        }
                    })
                    .setNegativeButton(android.R.string.cancel, null);

            return builder.create();
        }
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/settings/preference/ProxySettingDialog.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.settings.preference

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.CheckBox
import android.widget.EditText
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.webkit.WebViewProxy
import jp.hazuki.yuzubrowser.ui.preference.CustomDialogPreference
import jp.hazuki.yuzubrowser.ui.settings.AppPrefs

class ProxySettingDialog @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : CustomDialogPreference(context, attrs) {
    private var mSaveSettings = true

    fun setSaveSettings(save: Boolean): ProxySettingDialog {
        mSaveSettings = save
        return this
    }

    override fun crateCustomDialog(): CustomDialogPreference.CustomDialogFragment {
        return SettingDialog.newInstance(mSaveSettings)
    }

    class SettingDialog : CustomDialogPreference.CustomDialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val view = LayoutInflater.from(context).inflate(R.layout.proxy_dialog, null)
            val checkBox = view.findViewById<CheckBox>(R.id.checkBox)
            val editText = view.findViewById<EditText>(R.id.editText)
            val httpsCheckBox = view.findViewById<CheckBox>(R.id.httpsCheckBox)
            val httpsText = view.findViewById<EditText>(R.id.httpsEditText)

            httpsCheckBox.setOnCheckedChangeListener { _, isChecked -> httpsText.isEnabled = isChecked }

            checkBox.isChecked = AppPrefs.proxy_set.get()
            editText.setText(AppPrefs.proxy_address.get())
            httpsCheckBox.isChecked = AppPrefs.proxy_https_set.get()
            httpsText.setText(AppPrefs.proxy_https_address.get())
            httpsText.isEnabled = httpsCheckBox.isChecked

            val builder = AlertDialog.Builder(activity)

            builder
                    .setView(view)
                    .setTitle(R.string.pref_proxy_settings)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        val context = context
                        val enable = checkBox.isChecked
                        val proxyAddress = editText.text.toString()
                        val enableHttps = httpsCheckBox.isChecked
                        val httpsProxyAddress = httpsText.text.toString()

                        WebViewProxy.setProxy(context, enable, proxyAddress, enableHttps, httpsProxyAddress)

                        if (arguments!!.getBoolean(SAVE)) {
                            AppPrefs.proxy_set.set(enable)
                            AppPrefs.proxy_address.set(proxyAddress)
                            AppPrefs.proxy_https_set.set(enableHttps)
                            AppPrefs.proxy_https_address.set(httpsProxyAddress)
                            AppPrefs.commit(context, AppPrefs.proxy_set, AppPrefs.proxy_address,
                                AppPrefs.proxy_https_set, AppPrefs.proxy_https_address)
                        }
                    }
                    .setNegativeButton(android.R.string.cancel, null)

            return builder.create()
        }

        companion object {
            private const val SAVE = "save"

            fun newInstance(save: Boolean): SettingDialog {
                val dialog = SettingDialog()
                val bundle = Bundle()
                bundle.putBoolean(SAVE, save)
                dialog.arguments = bundle
                return dialog
            }
        }
    }
}
